Softmax improvement

bordesf opened the pull request:
Ref to #5712 

- Some modifications to Softmax and GradSoftmax to allow a N-dimensional array as input and apply the softmax operation on the last dimension.

For the next commits, I will need to add and change some tests in test_nnet.py. With the new implementation if the input is a vector, the output will be a vector. Currently the tests consider the fact that the softmax give a matrix if the input is a vector, so the test will crash in those cases.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -423,20 +423,17 @@ class Softmax(gof.Op):
-        x = tensor.as_tensor_variable(x)
-        if x.type.ndim not in (1, 2) \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
```
You should still raise an error if the dtype is not in ```float_dtypes ```.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -423,20 +423,17 @@ class Softmax(gof.Op):
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
-                             x.type)
+        x = theano.tensor.as_tensor_variable(x)
```
```tensor``` is already imported. So change this line to ```x = tensor.as_tensor_variable(x)```

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,39 +301,28 @@ class SoftmaxGrad(gof.Op):
         dy = tensor.as_tensor_variable(dy)
         sm = tensor.as_tensor_variable(sm)
-        if dy.type.ndim not in (1, 2) \
-                or dy.type.dtype not in tensor.float_dtypes:
```
You should still raise an error if the dtype is not in ```float_dtypes ```.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,39 +301,28 @@ class SoftmaxGrad(gof.Op):
-        if dy.ndim == 1:
-            dy = tensor.shape_padleft(dy, n_ones=1)
-        if sm.ndim == 1:
-            sm = tensor.shape_padleft(sm, n_ones=1)
```
Keep lines 309-312. We still need to handle vectors.

bordesf replied:
The goal is to handle vector without the need to pad them. (We don't really need to transform them into a matrix).

lamblin replied:
Some tests would assume that the softmax of a vector is a row, so that `softmax(v)[arange(1), y]` still inserts the `cross_entropy_softmax...` Op.
It is OK to change it in this PR, but we will have to handle the transition.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,39 +301,28 @@ class SoftmaxGrad(gof.Op):
     def perform(self, node, input_storage, output_storage):
         dy, sm = input_storage
+        if (dy.shape != sm.shape):
+            raise ValueError('dy and the softmax output should have the same shape.')
```
Isn't this always guaranteed? @lamblin  What do you think?

lamblin replied:
Yes, the shape of a variable and of the gradient wrt that variable should have the same shape.

ReyhaneAskari replied:
So, please remove lines 310 and 311.

bordesf replied:
In the old version of the code (The current one of Theano), they check if dy and sm has the same shape in the C code. So, to be consistent with the C code, I though it will be needed to do the same things in the python code.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -465,119 +462,114 @@ def c_code_template(dtype):
-                     "failed to alloc sm output");
-                %(fail)s
+            // Allocate memory for the softmax output
+            if (NULL == %(sm)s || (PyArray_DIMS(%(x)s) != PyArray_DIMS(%(sm)s)))
```
@lamblin. Do we need to check each dimension separately or we can check it like that. I can't think of a case where ```x.shape != sm.shape```. Could you give an example? Then we can also add it to the tests.

lamblin replied:
We need to check each dimension separately (I had not seen your comment here before commenting earlier about that).
It is possible that output memory `%(sm)s` is kept around between different calls to the compiled theano function, but that the shape of the arguments changed. We re-use it only if it has the right shape.
(We could try to be more clever, and re-use it if it is too big, but not too small, but it brings additional issues.)
DebugMode with `check_preallocated_out` checks this sort of thing, but I don't think it has been run in a while (it is really slow).

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -465,119 +462,114 @@ def c_code_template(dtype):
             {
-                dtype_%(sm)s row_ij = x_i[j * Sx1] ;
-                //std::cout << "2 " << j << " " << row_ij << " " << row_max << "\\n";
+                dtype_%(sm)s row_ij = x_i[j] ;
```
You need to use the stride here to move in a row: 

```
dtype_%(sm)s row_ij = x_i[j * stride_of_the_last_dimentsion]
```

bordesf replied:
In this case, we are on the last dimension. We don't really need to have a stride here since it's value will be always 1. But I agree in other cases, we should compute it.

lamblin replied:
It will not always be 1. The arrays are not guaranteed to be contiguous, even though they often are.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -465,119 +462,114 @@ def c_code_template(dtype):
-                //std::cout << "3 " << j << " " << sm_ij << "\\n";
                 sum += sm_ij;
-                sm_i[j * Ssm1] = sm_ij;
+                sm_i[j] = sm_ij;
```
You need to use the stride here to move in a row: 

```
sm_i[ j  * stride_of_the_last_dimentsion ] = sm_ij;
```

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -465,119 +462,114 @@ def c_code_template(dtype):
-            dtype_%(sm) s* __restrict__ sm_i = (dtype_%(sm)s*)(PyArray_BYTES(%(sm)s) + PyArray_STRIDES(%(sm)s)[0] * i);
+            // Compute stride
+            npy_int stride;
+            stride = PyArray_DIM(%(x)s, ndim - 1) * PyArray_ITEMSIZE(%(x)s);
```
Have you tested this? c_code is sometimes loaded from the cache and older versions were used. 

@lamblin for the ```c_code ``` function we needed to clean the cache or disable the c_code_cache. How can Florian do it in this  ```c_code_template``` function?

bordesf replied:
When I was testing this, I just disable the c_code_cache. I just put it back when I committed.

lamblin commented:
I still have not gotten around to reviewing this, I'll try to do that tomorrow.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,39 +301,28 @@ class SoftmaxGrad(gof.Op):
-            dy_times_sm_i = dy[i] * sm[i]
-            dx[i] = dy_times_sm_i - sum(dy_times_sm_i) * sm[i]
+        dy_times_sm = dy * sm
+        dx = dy_times_sm - (numpy.sum(dy_times_sm, axis=-1, keepdims=True) * sm)
```
OK. I'd still keep the comment with the math (or update it).

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,39 +301,28 @@ class SoftmaxGrad(gof.Op):
 
     def infer_shape(self, node, shape):
-        return [shape[1]]
+        return [shape[-1]]
```
`shape[1]` here means the shape of the symbolic input number 1 (here, `sm` as passed to `make_node`).
it works with `[-1]` as well, but its less clear.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,39 +301,28 @@ class SoftmaxGrad(gof.Op):
+        return [shape[-1]]
 
     def c_code_cache_version(self):
         return (3,)
```
Update the version number if you update the C code

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
+                    %(fail)s;
+                }
+
+            // Check input type
```
Maybe put that comment before the first check.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
+                }
+
+            // Check the shapes
+            if (PyArray_DIMS(%(dy)s) == PyArray_DIMS(%(sm)s))
```
PyArray_DIMS just returns a pointer to the array of shapes, the comparison will fail.
You will need to loop and compare each `PyArray_DIMS(...)[i]`.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
+                }
+
+            // Alloc memory for the output
+            if ((NULL == %(dx)s) || (PyArray_DIMS(%(dx)s) != PyArray_DIMS(%(sm)s)))
```
Idem here, you will have to loop to compare the shapes.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
+
+            // Compute array size except on the last dimension
+            npy_int input_size;
+            input_size = PyArray_SIZE(%(sm)s) / input_shape[ndim - 1];
```
This could crash the program if `input_shape[ndim - 1]` is equal to 0.
Could you detect that case and raise a Python exception instead? It makes things much easier to debug.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
+
+            // Compute stride softmax
+            npy_int stride;
+            stride = PyArray_DIM(%(sm)s, ndim - 1) * PyArray_ITEMSIZE(%(sm)s);
```
It's more complex than that if the arrays are not contiguous. The stride can also be different for the different arrays.
That being said, I believe numpy gives iterators that can be easier to use than trying to compute the strides explicitly.

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
+                double sum_dy_times_sm = 0.;
+                for (size_t j = 0; j < input_shape[ndim - 1]; ++j)
+                {
+                    dx_i[j] = dy_i[j] * sm_i[j];
```
`dx_i`, `dy_i`, and `sm_i` can all have different strides as well...

bordesf commented:
Should I use the vectorized version of exp in logSoftmax if is available ? (Like in Softmax op)

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,109 +301,151 @@ class SoftmaxGrad(gof.Op):
+            npy_intp* input_shape = PyArray_DIMS(%(sm)s);
+            npy_int ndim_sm = PyArray_NDIM(%(sm)s);
+            npy_int ndim_dy = PyArray_NDIM(%(dy)s);
+            npy_intp* shape_sm = PyArray_DIMS(%(sm)s);
```
Lines 337 and 340 are the same. Remove line 337 and replace ```input_shape``` with ```shape_sm```, in the rest of the code.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -302,109 +301,151 @@ class SoftmaxGrad(gof.Op):
+                    %(fail)s;
+                }
+
+            // Check the shapes of sm and dy
```
@pascal, I presume that shape of ```dy``` and ```sm``` are always the same. Right? So I don't understand why such a check has been here

ReyhaneAskari replied:
@Florian, if needed, you also need to check if ```dx``` has the same shape as ```sm```.

bordesf replied:
I already check that few lines below (Between 384 and 393).

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -645,79 +701,98 @@ def c_headers(self):
+            stride_x = PyArray_STRIDE(%(x)s, last_axis)/sizeof(dtype_%(x)s);;
+            stride_sm = PyArray_STRIDE(%(sm)s, last_axis)/sizeof(dtype_%(sm)s);
+            // Get the shape on the specified dimension
+            num_data = shape_x[last_axis];
```
Use a better name instead of ```num_data```. Maybe, ```num_sm_classes```

bordesf replied:
Yeah, it will be more meaningful, thanks

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -465,119 +503,138 @@ def c_code_template(dtype):
 
             inside_row_loop = """
-            if(Ssm1 == 1){
                 %(inside_row_loop_contig)s
```
What if it's not contiguous? Shouldn't you keep the else here? Or I'm missing where are you using
```inside_row_loop```?

bordesf replied:
I corrected that in a new version that I didn't committed yet.

bordesf commented:
I modified the code in order to select the axis on where we want to compute the softmax. I added an input node "axis" and I was wondering which method should I use about the gradient of this new input. For now, I used the DisconnectedType but maybe disconnected_grad or consider_constant would be better. What do you think ?

ReyhaneAskari commented:
I don’t know why you have chosen to have a node for the axis and not simply an integer. I presume we rarely need to change it on the run time. @lamblin what do you think?

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -54,35 +54,30 @@ class SoftmaxWithBias(gof.Op):
-        if x.type.ndim != 2 \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 2-d tensor of floats')
+        axis = tensor.as_tensor_variable(axis)
```
if `axis` is supposed to be a scalar, you should also check that its number of dimensions is 0.

f0k replied:
The question is whether this should be symbolic at all. `T.sum()`, `T.max()` etc. have nonsymbolic `axis` arguments.

lamblin replied:
Non-symbolic arguments would be fine by me too.
It could make some things much easier, for instance shape inference.

f0k replied:
> It could make some things much easier, for instance shape inference.

... and optimization (we will want to use cuDNN if `axis==1` and `x.ndim in (4,5)`), and replacing it with an expression graph that uses `T.sum()` and `T.max()` (if ever the compiled expression graph becomes as fast as the manual C code).

lamblin commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,281 +127,372 @@ def c_code_template(dtype):
+                    }
+
+            // Check if the axis is valid
+            if(axis > ndim_x - 1 || axis < -1){
```
A negative axis means to use `ndim_x + axis` instead. It is not just a special case for `-1`, it works for all negative axes in numpy.
So the checks should be something like:
```C
if (axis < 0)
    axis += ndim_x;
if (axis > ndim_x - 1 || axis < 0)
    PyErr_SetString(...);
```

f0k commented:
Hmm, why do we need an abstract softmax? Can anybody enlighten me? Is it that implementing a Softmax Op supporting arbitrary axes is too complicated, and you'd rather add an optimizer inserting a dimshuffle->reshape->softmax->reshape->dimshuffle graph as needed?

ReyhaneAskari commented:
@f0k , well I think he has used the abstract softmax for the general case where we can have a list of axes to apply the softmax on. In that case, he uses the old softmax when the list has only one element. The purpose is to not lose the optimizations that already exist for the old softmax. 

Implementing an op that would  _dimshuffle->reshape->softmax->reshape->dimshuffle_ would have overhead. We now have the implementations in C code which is much faster.

f0k commented:
> I think he has used the abstract softmax for the general case where we can have a list of axes to apply the softmax on. In that case, he uses the old softmax when the list has only one element.
> [...]
> Implementing an op that would dimshuffle->reshape->softmax->reshape->dimshuffle would have overhead. We now have the implementations in C code which is much faster.

If the plan is to eventually implement the general multi-axis softmax in C, then I don't see the reason for the abstract/placeholder softmax. What is the advantage over extending the old softmax to support multiple axes?

ReyhaneAskari commented:
Implementing a softmax that would cover all the cases is very hard at this point. We would rather continue this approach and add the functionality to theano. @bordesf can probably explain better in more details.

f0k commented:
> Implementing a softmax that would cover all the cases is very hard at this point.

I agree. This will be complicated in C.

> We would rather continue this approach and add the functionality to theano.

So I'm back at assuming there will be an optimizer replacing the abstract softmax with a dimshuffle->reshape->(old)softmax->reshape->dimshuffle whenever the abstract softmax has multiple axes (and they don't coincide with what's supported by cuDNN softmax, which would be another optimizer)?

lamblin commented:
@f0k: the first option is correct: optimize to cuDNN if possible, otherwise dimshuffle / reshape if multiple axes, otherwise the regular Op if only one axis. We may add other specially-optimized cases later if needed.

bordesf commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,284 +131,574 @@ def c_code_template(dtype):
-                or dy.type.dtype not in tensor.float_dtypes:
-            raise ValueError('dy must be 1-d or 2-d tensor of floats. Got ',
-                             dy.type)
+        axis = tensor.as_tensor_variable(axis)
```
If we want to keep the axis as a node in the graph, we still need to define it as a tensor variable.
Yeah, I will do some cleaning in the code.

ReyhaneAskari replied:
Why do we want to have the axis as a node?

bordesf commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,329 +132,496 @@ def c_code_template(dtype):
+    def L_op(self, inp, outputs, grads):
+        x, axes = inp
+        # To investigate
+        if len(grads) == 2:
```
During the rop test, len of grads can be two and I don't really get why it happens, it should be one. Do you have an idea ? @ReyhaneAskari

ReyhaneAskari replied:
Well I don't really know but I presume that in the case of Softmax with Bias, ```grads[1]``` is wrt bias term . I think you can simply ignore it. Also, you don't need to define an ```if .. else``` here.  ```g_sm, = grads``` will always take the first element. @lamblin am I right?

f0k replied:
> `g_sm, = grads` will always take the first element.

No, it will complain when grads is not an iterable of size 1. `g_sm = grads[0]` would work for both cases. Or `g_sm, *_ = grads` in Python 3.

bordesf replied:
@f0k Yes, I will use `g_sm = grads[0]`, thanks.
@ReyhaneAskari I don't think it corresponds to a bias term because in the current Theano code,  we have `g_sm, = grads` and it means that grads is a list of length one (Otherwise as @f0k said, it will not work). I think, I have done something inside the optimization that add this term (Maybe it corresponds to an axis, when I run the rop.test_softmax tests, the second element of grads is `[Elemwise{second,no_inplace}.0]` that corresponds to `[0]`. I would like to understand what's going on. `g_sm, = grads[0]` works but I am wondering if this second element in grads can have any kind of importance.

f0k replied:
> Maybe it corresponds to an axis

@bordesf: I haven't fully followed the code, but based on your comment, is the axis still registered as an input node for the softmax op? Wouldn't it be easier to make it a nonsymbolic attribute? We don't want the axis/axes to be changed at runtime anyway.

bordesf replied:
I spoke with @lamblin and as consequences of having axis as a node, we get the grad of the axis in eval_points, since this grad isn't defined we don't need to pass it through the L_op. So I put as argument eval_point[0] and it resolves the issue.
Yes, we keep the axis as input node. It simplifies the access of the axis value during the optimization. I think @lamblin can give other reasons about this choice.

lamblin replied:
Honestly, both could work. Having it an input will generate fewer versions of the C code, but that could be mitigated by making it a Param instead. I'll try to see with @notoraptor if that would be a better solution.

bordesf commented:
I add a PR #5980 where I introduce a function that normalize the axes to avoid some code redundancy.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -93,23 +93,25 @@ def perform(self, node, input_storage, output_storage):
-        e_x = np.exp(x_plus_b - x_plus_b.max(axis=1)[:, None])
-        e_x *= 1.0 / e_x.sum(axis=1)[:, None]
+        # We add bias on the last dimension
+        x_plus_b = x + b
```
Hi @bordesf ,

Here, you cannot sum x and b as they don't have the same shape. You need to make it broadcastable first. Also you need to change the C code accordingly.
A test case:
```python
a = T.tensor3('a')
b = T.vector('b')
res =  softmax_with_bias(a, b, 0)
f = theano.function([a, b], res)
a_ = np.random.rand(3, 4, 5)
b_ = np.random.rand([3, 4, 5][axis])
res_ = f(a_, b_)
```
```
ValueError: operands could not be broadcast together with shapes (3,4,5) (3,) 
Apply node that caused the error: SoftmaxWithBias(a, b, TensorConstant{0})
Toposort index: 0
Inputs types: [TensorType(float64, 3D), TensorType(float64, vector), TensorType(int8, scalar)]
Inputs shapes: [(3, 4, 5), (3,), ()]
Inputs strides: [(160, 40, 8), (8,), ()]
Inputs values: ['not shown', array([ 0.66083873,  0.16459271,  0.09435837]), array(0, dtype=int8)]
Outputs clients: [['output']]```

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -800,15 +1143,94 @@ def softmax_graph(c):
+        # If axis is None, we select the last axis
+        if NoneConst.equals(axis):
+            axis = -1
+        elif not isinstance(axis, tensor.var.TensorConstant):
```
There is no ```tensor.var.TensorConstant```. You probably meant ```tensor.TensorConstant```?

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -800,15 +1143,94 @@ def softmax_graph(c):
+                    (isinstance(axis.data, np.ndarray) and axis.data.ndim == 0):
+                        axis = [int(axis.data)]
+            elif isinstance(axis.data, (list, np.ndarray)):
+                axis = [int(i) for i in axis.data]
```
Why not simply ```axis = axis.data```? Why the casting is needed?

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -800,15 +1143,94 @@ def softmax_graph(c):
+        # If axis is None, we select the last axis
+        if NoneConst.equals(axis):
+            axis = -1
+        elif not isinstance(axis, tensor.var.TensorConstant):
```
The same issue here.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,329 +132,486 @@ def c_code_template(dtype):
-        sm = e_x / e_x.sum(axis=1)[:, None]
+        x, axes = input_storage
+        # Apply softmax on the specified dimension
+        e_x = np.exp(x - x.max(axis=axes, keepdims=True))
```
Try this example:
```python
a = T.tensor3('a')
axis = T.constant([0, 1, -1])
res = softmax(a, axis=axis)
f = theano.function([a], res)
a_ = np.random.rand(3, 4, 5)
res_ = f(a_)
print res_.shape
```
When we reach here, ```axes``` is ```array([0, 1, 2])``` which leads to an error. A fix that comes to my mind is to cast it to tuple like ```x.max(axis=tuple(axes), keepdims=True)```.

ReyhaneAskari replied:
@bordesf , any updates? This is still failing.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,329 +132,486 @@ def c_code_template(dtype):
+        x, axes = input_storage
+        # Apply softmax on the specified dimension
+        e_x = np.exp(x - x.max(axis=axes, keepdims=True))
+        sm = e_x / e_x.sum(axis=axes, keepdims=True)
```
The same here.

ReyhaneAskari replied:
Still failing.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -577,57 +773,57 @@ def c_code_template(dtype):
-                            keepdims=True))
+        x, axes = input_storage
+        xdev = x - x.max(axis=axes, keepdims=True)
+        lsm = xdev - np.log(np.sum(np.exp(xdev), axis=axes, keepdims=True))
```
The problem with tuple here as well.

bordesf commented:
@ReyhaneAskari Thanks for your review, I push some commits to correct what you mentioned. About the axes, I used the function of PR #5980. A todo for the next days, is to raise an error in case of the len of axis is superior to one when using the old softmax.

f0k commented on theano/tensor/nnet/nnet.py:
```diff
@@ -401,6 +396,11 @@ def make_node(self, dy, sm, axis=-1):
+            opt.get_scalar_constant_value(axis)
+        except tensor.NotScalarConstantError:
+            raise ValueError('axis must be a scalar. Use abstract_softmax instead Got ', axis.type)
         return Apply(self, [dy, sm, axis], [sm.type()])
```
Sorry, but I still don't get why `axis` is a symbolic input node at all. `theano.tensor.elemwise.Sum` has a nonsymbolic tuple of axes:
```
>>> import theano.tensor as T
>>> x = T.matrix('x')
>>> y = x.sum(axis=-1)
>>> y.owner.op.axis
(1,)
>>> y.owner.inputs
[x]
```
What's the advantage of making it symbolic?

bordesf replied:
If there are no advantages of making it symbolic, it makes sens that it should become nonsymbolic. I will wait the answer of @lamblin and @notoraptor before making new commits.

bordesf replied:
After discussion, we agreed to use a params type: http://deeplearning.net/software/theano_versions/dev/library/gof/params_type.html instead of the symbolic axis.

ReyhaneAskari commented on theano/tensor/basic.py:
```diff
@@ -968,6 +968,53 @@ def _pack(x):
+        axis = []
+    elif (isinstance(axis, (integer_types, np.integer)) or
+            (isinstance(axis, np.ndarray) and axis.ndim == 0)):
+                axis = [int(axis)]
```
I know this is merged but you need to fix a corner case. If the ```axis=np.array(1.5)``` it will be floored to 1 while we should raise an error.

ReyhaneAskari commented on theano/tensor/basic.py:
```diff
@@ -968,6 +968,53 @@ def _pack(x):
+    elif (isinstance(axis, (integer_types, np.integer)) or
+            (isinstance(axis, np.ndarray) and axis.ndim == 0)):
+                axis = [int(axis)]
+    elif isinstance(axis, (tuple, list, np.ndarray)):
```
Check the type here as well to be int.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,329 +130,496 @@ def c_code_template(dtype):
+    nout = 1
+    __props__ = ()
+
+    def make_node(self, x, axis=-1):
```
At this point we already processed the axis and it should be a list of valid axes. So no default value is needed.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,329 +130,496 @@ def c_code_template(dtype):
+        if x.ndim == 1:
+            x = tensor.shape_padleft(x, n_ones=1)
+        # Check if axis is an integer
+        if axis.type.dtype not in tensor.int_dtypes:
```
This check is also already done in check_and_normalize_axis. Remove this.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -130,329 +130,496 @@ def c_code_template(dtype):
+        # Check if axis is a scalar
+        try:
+            opt.get_scalar_constant_value(axis)
+        except tensor.NotScalarConstantError:
```
I feel like this check is unnecessary. Again the reason is that axis is already processed in check_and_normalize_axis. Am I missing something? Could you come up with a case that check_and_normalize_axis passes but here the axis variable is not scalar? If so, I presume this check must be moved to check_and_normalize_axis.

I appreciate it if you reply to my comments whenever you fix them by a new commit. I want to make sure that you've seen these.

bordesf replied:
check_and_normalize_axis return a list of sorted integers. This Softmax op with the C code optimization corresponds to the old op that manages only the case where we specify **only one axis**. So, if check_and_normalize_axis return (2,3) and this op is called instead of the Abstract_Softmax, it will fail because the C code doesn't manage more than one axis. I think we should really keep it because in case where there are some issue in the optimization it will be much easier to detect the case where a wrong op is called.

ReyhaneAskari replied:
Thanks for the explanation. Maybe we could add a comment here that explains what you expect.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -2445,7 +2815,7 @@ def elu(x, alpha=1):
         Exponential Linear Units (ELUs)" <http://arxiv.org/abs/1511.07289>`.
     """
-    return tensor.switch(x > 0, x, alpha * tensor.expm1(x))
+    return tensor.switch(x > 0, x, alpha * (tensor.exp(x) - 1))
```
For small values of x, tensor.expm1(x) provides higher precision. So it's better not to remove this.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -1435,8 +1804,10 @@ def make_node(self, coding_dist, true_one_of_n):
+            _coding_dist = tensor.shape_padleft(_coding_dist, n_ones=1)
         if _coding_dist.type.ndim != 2:
-            raise TypeError('matrix required for argument: coding_dist')
+            raise TypeError('iatrix required for argument: coding_dist')
```
Typo.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -800,15 +1155,20 @@ def softmax_graph(c):
-    return logsoftmax_op(c)
+def logsoftmax(c, axis=(-1,)):
+    c = tensor.as_tensor_variable(c)
+    if c.broadcastable[-1]:
```
The same here.

ReyhaneAskari commented on theano/tensor/nnet/nnet.py:
```diff
@@ -637,108 +841,258 @@ def R_op(self, inputs, eval_points):
+        x = tensor.as_tensor_variable(x)
+        axis = tensor.as_tensor_variable(axis)
+        # Check if axis is an integer
+        if axis.type.dtype not in tensor.int_dtypes:
```
We checked these already. See my other comment.

ReyhaneAskari commented:
@bordesf I did the minor fixes so that you can focus on replacing the symbolic nodes by param types and @shawntan  can start working on the tests.

bordesf commented:
Since this pull request implies too much modifications of the code, after discussion with @lamblin and @ReyhaneAskari, in order to improve the readability and to make the review easier, I will create different pull requests with few features for each.

ReyhaneAskari commented:
thanks @bordesf. Please tag me in your new PRs.
